src/main/groovy/com/google/protobuf/gradle/ToolsLocator.groovy

```diff
@@ -104,8 +105,8 @@ class ToolsLocator {
       group:groupId,
       name:artifact,
       version:version,
-      classifier:classifier ?: osdetector.classifier,
-      ext:extension ?: 'exe',
+      classifier:classifier ?: mainClass ? null : osdetector.classifier,
```

Comment:
mainClass should virtually never be specified. In your specific case for dubbo it makes some sense, as the JAR has multiple generators inside. But the recommendation is "the JAR should be stand-alone, with main-class specified in the meta-inf". So I doubt we want these two lines of convenience, as it creates a perverse incentive.

Reply:
If mainClass can be specified, it will be more convenient to use.

Reply:
I am currently in urgent need of this feature. Without it, I may need to maintain multiple jars separately to complete it, and others may also need to find other ways to solve this problem

Reply:
I do not want to encourage mainClass; most cases there should only be a single class. Allowing alternative classes and encouraging never having Main-Class in the jar are two very different things.

Reply:
![image](https://github.com/google/protobuf-gradle-plugin/assets/55786929/aae8a0c7-48d0-498d-9d55-6d6681241b46)
![image](https://github.com/google/protobuf-gradle-plugin/assets/55786929/2983ada6-d467-4254-a01a-961d4f009424)

How should I specify classifier as null

Reply:
If alternative classes are allowed, notification.ext needs to be a jar, and notation.classifier also needs to be null, otherwise it will be replaced by osdetector.classifier, resulting in the jar being unable to be found.

Reply:
I used translation software which may not convey the message accurately.

Reply:
The problem isn't limited to "allowing alternative classes to be the main class." The problem you're seeing looks to impact all usages of jars. My point here is I don't want behavior tied to overriding the main class. Whatever solution here is needed should also be available when using the main-class from inside the jar.

Taking a look at ServiceTalk, it looks like they are just specifying a classifier (`io.servicetalk:servicetalk-grpc-protoc:$serviceTalkVersion:all@jar`). They were the ones that added jar support to the plugin, so I guess they just didn't need null classifier. Since the gradle shadow plugin does encourage you to publish both "jar without dependencies" and the fat jar, I guess that makes some sense.

Do you only need dubbo-compiler support? It looks like their jar doesn't include dependencies, so I don't think the null classifier would have worked. I'm impressed that Dubbo has gone out of their way to make the exe's.

Reply:
![image](https://github.com/google/protobuf-gradle-plugin/assets/55786929/3a8bdb99-8106-48db-ac8d-8f7f3d1344e6)
![image](https://github.com/google/protobuf-gradle-plugin/assets/55786929/a4a09039-603a-46fc-8764-a9ec75cbd099)
![image](https://github.com/google/protobuf-gradle-plugin/assets/55786929/5d28d368-66c6-4b8c-8ff7-5172c0215838)
If I want to use a jar, but this jar does not have a classifier, it will be forced to specify a classifier by osdetector.classifier. In fact, many jars do not have a classifier. In this case, ar cannot be selected at all.


`classifier:classifier ?: mainClass ? null : osdetector.classifier,`
Actually, I want jars that do not need classifier to not be overwritten by osdetector.classifier

Reply:
I understand theoretically wanting a `null` classifier. But specifically for dubbo-compiler it seems to be pointless, as dubbo-compiler isn't a fat jar. What plugin are you wanting to use here? Is it dubbo-compiler, or something else?

Reply:
Yes, I want to use dubbo compiler. I have recompiled a complete jar in my private library